### PR TITLE
Hide items with no buy history until listed for first time

### PR DESCRIPTION
**Pull request by CatsEyeXI**
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [X] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [X] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

This code was written by @OpheliaXI. I am simply porting it over to LSB.

## What does this pull request do?
This PR will hide any items from the Auction House that do not have a buy history.
This is a more realistic scenario for servers, as some gear is unobtainable or uncoded.
If the operator chooses, they can use a 3rd party tool like pydarkstar to populate their auction house.

## Steps to test these changes
Before change:
You will see every single item, whether it's been purchased before or not

After change:
Only items with a buy history are visible on the auction house.

### CODE REVIEW

**Comment by zach2good** [comment]
I'm not against the idea, but it isn't 'as retail', since retail doesn't have any reason to omit any results.

You could wrap the whole additional block in a setting in search.lua and use that to toggle the new behaviour, defaulting to false. OMIT_NO_HISTORY or something

**Comment by CatsEyeXI** [comment]
> I'm not against the idea, but it isn't 'as retail', since retail doesn't have any reason to omit any results.
> 
> You could wrap the whole additional block in a setting in search.lua and use that to toggle the new behaviour, defaulting to false. OMIT_NO_HISTORY or something

I think I did that right, seems to work.

**Comment by TeoTwawki** [comment]
What retail does is track if the item has ever been _listed_ before. even if it never sells, as long as somebody made the attempt its in the list forever after that,

**Review by TeoTwawki — APPROVED** [review]
if you ask me, we're better than retail. and this setting is the closest we're getting to what retail does without a more painful reworking. So I aprove.

File: src/search/data_loader.cpp [hunk 1]
```diff
@@ -119,6 +119,49 @@ std::vector<ahItem*> CDataLoader::GetAHItemsToCategory(uint8 AHCategoryID, int8*
...
+                uint16 CurrItemID = sql->GetUIntData(0);
+                bool   itemFound  = false;
+                for (ahItem* PAHItem : ItemList)
+                {
+                    if (PAHItem->ItemID == CurrItemID)
+                    {
+                        itemFound = true;
+                        break;
+                    }
+                }
```
Comment (TeoTwawki) [inline src/search/data_loader.cpp:148]: failed to notice the lack of line break here before I reflex-approved.

**Review by TeoTwawki — CHANGES_REQUESTED** [review]

**Comment by 0phie** [comment]
So we would need a separate table to get this to operate the exact way Teo described retail operating as. This as although putting it in item_basic.sql would be great, it would likely lead to overwriting of that data pretty quick. It wouldn't be hard to add a new protected table that we check when an item is listed with a "Hey this has been listed before" or "This is a new item to the AH".

**Comment by TeoTwawki** [comment]
> So we would need a separate table to get this to operate the exact way Teo described retail operating as. This as although putting it in item_basic.sql would be great, it would likely lead to overwriting of that data pretty quick. It wouldn't be hard to add a new protected table that we check when an item is listed with a "Hey this has been listed before" or "This is a new item to the AH".

I'm not sure anyone in or out of the team cares about that feature enough to make it 100% that way in the 1st place.

Item basic is read only atm, so I don't think we'd want to store there. We could query the AHs rows to determine if the item was ever _listed_ (a row will exist, even if it hasn't sold, if we don't discard that row on return - some adjusting would be required) but honestly a new table to store some data like "have we seen this before" seems a lot saner - but really only if we also stopped doing the whole "lets load everything from item basic into here!" that we currently do. And really in a lot of ways the AH is what I call better-than-retail, so its kind of awkward to hold to the "must be retail accurate" standard 100%

I can actually prove what retail does there tho, if anyone cares. :)

I think this is fine as is till someone just can't stand that we're imperfect.

File: src/search/data_loader.cpp (hunk 1 above)
Comment (CatsEyeXI) [reply src/search/data_loader.cpp:148]: Hi Teo, I'm not clear on what change I need to make here, can you please explain? :)

Comment (TeoTwawki) [reply src/search/data_loader.cpp:148]: new line between the closing brace and the if that follows. not sure why style checks didn't complain.

Comment (TeoTwawki) [reply src/search/data_loader.cpp:148]:
```
}
if (stuff)
{
    moreStuff();
}
```
vs 
```
}

if (stuff)
{
    moreStuff();
}
```

**Review by TeoTwawki — APPROVED** [review]

**Review by zach2good — APPROVED** [review]

**Pull request merged**